### PR TITLE
A few fixes and multicluster observables

### PR DESCRIPTION
**Pull request by beaudett**
Fixed bugs : 1) reachedEE which had been turned into an int (1=barrel, 2=endcaps)  2) the extrapolated positions were no longer stored. 
Added features: implementation of PCA analysis to compute the position and the axis of the shower. Added some multicluster observables: number of layers, first layer, last layer, sigma_uu and vv which are the equivalent of sigma-eta-eta and sigma-phi-phi but in centimeters. 
In the PCA computation, a weight based on the mip value is introduced, which required modifying the config file.

Apologies for the differences in the indentation introduced by my editor that I have tried to undo as much as possible.

### CODE REVIEW

**Comment by beaudett** [comment]
Wait a bit before merging this PR; I think I found a bug.

**Comment by beaudett** [comment]
Ok. The bug was in my test. You can go ahead

File: HGCalAnalysis/test/exampleConfig.py [hunk 1]
```diff
@@ -10,18 +10,17 @@
...
 from FastSimulation.Event.ParticleFilter_cfi import *
+from RecoLocalCalo.HGCalRecProducers.HGCalRecHit_cfi import dEdX_weights as dEdX
 
 process.maxEvents = cms.untracked.PSet( input = cms.untracked.int32(-1) )
 
 process.source = cms.Source("PoolSource",
     # replace 'myfile.root' with the source file you want to use
     fileNames = cms.untracked.vstring(
-        '/store/relval/CMSSW_9_1_0_pre3/RelValTTbar_14TeV/GEN-SIM-RECO/PU25ns_91X_upgrade2023_realistic_v1_D13PU200-v2/10000/04A22787-5E31-E711-A724-0025905A6090.root'
+        '/store/relval/CMSSW_9_1_0_pre1/RelValTTbar_14TeV/GEN-SIM-RECO/PU25ns_90X_upgrade2023_realistic_v9_D4TPU200r3-v1/00000/02E3478F-3D22-E711-80C4-0CC47A4D7654.root'
```
Comment (clelange) [inline HGCalAnalysis/test/exampleConfig.py]: Can you undo the change for this line? The geometry of this file does not correspond to the one used in the config.

File: HGCalAnalysis/plugins/HGCalAnalysis.cc [hunk 2]
```diff
@@ -386,6 +424,27 @@ HGCalAnalysis::HGCalAnalysis(const edm::ParameterSet& iConfig) :
...
+	t->Branch("multiclus_eigenVal2", &multiclus_eigenVal2);
+	t->Branch("multiclus_eigenVal3", &multiclus_eigenVal3);
+	t->Branch("multiclus_eigenSig1", &multiclus_eigenSig1);
+	t->Branch("multiclus_eigenSig2", &multiclus_eigenSig2);
+	t->Branch("multiclus_eigenSig3", &multiclus_eigenSig3);
+	t->Branch("multiclus_siguu", &multiclus_siguu);
+	t->Branch("multiclus_sigvv", &multiclus_sigvv);
+	// t->Branch("multiclus_sigrad", &multiclus_sigrad);
+	// t->Branch("multiclus_radsiguu", &multiclus_radsiguu);
+	// t->Branch("multiclus_radsigvv", &multiclus_radsigvv);
```
Comment (clelange) [inline HGCalAnalysis/plugins/HGCalAnalysis.cc]: Do you intend to use these variables? If not, would be better to remove them completely (also below)

File: HGCalAnalysis/plugins/HGCalAnalysis.cc [hunk 3]
```diff
@@ -836,7 +921,75 @@ HGCalAnalysis::analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup)
 		multiclus_slopeY.push_back(multiClusters[i].y());
 		multiclus_cluster2d.push_back(cl2dIndices);
 		multiclus_cl2dSeed.push_back(cl2dSeed);
+		multiclus_firstLay.push_back(*layers.begin());
+		multiclus_lastLay.push_back(*layers.rbegin());
+		multiclus_NLay.push_back(layers.size());
+		if(multiClusters[i].size()<3)
+		  	{
+			multiclus_pcaAxisX.push_back(-2.);
```
Comment (clelange) [inline HGCalAnalysis/plugins/HGCalAnalysis.cc]: is -2 a good default value, i.e. will this never happen in reality for the new variables? Maybe use -999?

File: HGCalAnalysis/plugins/HGCalAnalysis.cc [hunk 4]
```diff
@@ -270,7 +305,10 @@ HGCalAnalysis::HGCalAnalysis(const edm::ParameterSet& iConfig) :
 	propagationPtThreshold(iConfig.getUntrackedParameter<double>("propagationPtThreshold",3.0)),
 	detector(iConfig.getParameter<std::string >("detector")),
 	rawRecHits(iConfig.getParameter<bool>("rawRecHits")),
-	particleFilter(iConfig.getParameter<edm::ParameterSet>("TestParticleFilter"))
+	particleFilter(iConfig.getParameter<edm::ParameterSet>("TestParticleFilter")),
+	dEdXWeights(iConfig.getParameter<std::vector<double> >("dEdXWeights")),
+	invThicknessCorrection({1./1.132,1./1.092,1./1.084}),
```
Comment (clelange) [inline HGCalAnalysis/plugins/HGCalAnalysis.cc]: This thickness correction can also be taken from the CMSSW configs?

File: HGCalAnalysis/plugins/HGCalAnalysis.cc (hunk 3 above)
Comment (beaudett) [inline HGCalAnalysis/plugins/HGCalAnalysis.cc]: -2 has the merit that one get a decent scale when TBrowsing it while -999 make it a "two-steps" process. I would like to keep -2 if you don't mind.

File: HGCalAnalysis/plugins/HGCalAnalysis.cc (hunk 2 above)
Comment (beaudett) [inline HGCalAnalysis/plugins/HGCalAnalysis.cc]: Will remove.

File: HGCalAnalysis/plugins/HGCalAnalysis.cc (hunk 4 above)
Comment (beaudett) [inline HGCalAnalysis/plugins/HGCalAnalysis.cc]: The parameters are here:
https://github.com/cms-sw/cmssw/blob/CMSSW_9_0_X/RecoLocalCalo/HGCalRecProducers/python/HGCalRecHit_cfi.py#L6-L58
I don't know if they be extracted from the definition of a producer. If there are put apart as done for dEdX_weights that would be ideal.

File: HGCalAnalysis/test/exampleConfig.py (hunk 1 above)
Comment (beaudett) [inline HGCalAnalysis/test/exampleConfig.py]: Ok. Will do.

File: HGCalAnalysis/plugins/HGCalAnalysis.cc (hunk 3 above)
Comment (beaudett) [inline HGCalAnalysis/plugins/HGCalAnalysis.cc]: this value is between -1 and 1. Setting -2 is convenient when checking the variable in a TBrowser

File: HGCalAnalysis/plugins/HGCalAnalysis.cc (hunk 4 above)
Comment (beaudett) [inline HGCalAnalysis/plugins/HGCalAnalysis.cc]: The URL was wrong, it is actually here
https://github.com/cms-sw/cmssw/blob/CMSSW_9_0_X/RecoLocalCalo/HGCalRecProducers/python/HGCalRecHit_cfi.py#L82
I searched a bit and I couldn't find a way to extract the parameters from a producer. I can modify HGCalRecHit_cfi.py to put these parameters separately if you so wish. There will be a delay of course.

Comment (beaudett) [inline HGCalAnalysis/plugins/HGCalAnalysis.cc]: I have prepared the modification in HGCalRecHit_cfi so that these parameters can be extracted. It is in 
https://github.com/beaudett/cmssw/tree/topic_extract_parameters_3
That would need to go first in CMSSW before we can use it in reco-ntuples.

File: HGCalAnalysis/plugins/HGCalAnalysis.cc (hunk 3 above)
Comment (artlbv) [inline HGCalAnalysis/plugins/HGCalAnalysis.cc]: I reran this branch on top of @clelange latest PR (https://github.com/cms-sw/cmssw/pull/19149) and now there seem to be no multiclusters with less than 2dclusters or layers. Therefore, this default value could be omitted.
Alternatively, I would suggest to just skip, i.e. not store, multiclusters with less then 3 layers.

Comment (edjtscott) [inline HGCalAnalysis/plugins/HGCalAnalysis.cc]: For a multicluster to be stored there is a requirement on the number of 2D clusters it has, [here](https://github.com/cms-sw/cmssw/blob/master/RecoLocalCalo/HGCalRecAlgos/src/HGCalDepthPreClusterer.cc#L68). The [default value](https://github.com/cms-sw/cmssw/blob/master/RecoLocalCalo/HGCalRecProducers/python/hgcalLayerClusters_cfi.py#L21) is three, so one would only expect multiclusters with at least four 2D clusters. 

I sometimes alter this to keep all the multiclusters so keeping this default seems sensible to me. (As an aside, the minClusters label is a bit misleading since the cut is implemented as a strictly greater than - we should change it at some point).

File: HGCalAnalysis/plugins/HGCalAnalysis.cc (hunk 4 above)
Comment (clelange) [inline HGCalAnalysis/plugins/HGCalAnalysis.cc]: Hi Florian, did you make a PR for this in CMSSW? I think it would be useful.

File: HGCalAnalysis/plugins/HGCalAnalysis.cc (hunk 3 above)
Comment (clelange) [inline HGCalAnalysis/plugins/HGCalAnalysis.cc]: I also think keeping this value makes sense. However, I guess the PCA will mostly be used by Florian and Arthur (maybe Ed?), so what about just adding an *additional* configurable flag (default false) to either stores these variables in the ntuple or not to avoid blowing up the ntuple size unnecessarily?

File: HGCalAnalysis/plugins/HGCalAnalysis.cc (hunk 4 above)
Comment (beaudett) [inline HGCalAnalysis/plugins/HGCalAnalysis.cc]: Hello Clemens, 
no, I didn't because it doesn't merge while the change is trivial. As I am leaving on vacation tonight, I will take a fresh look in September.

File: HGCalAnalysis/plugins/HGCalAnalysis.cc (hunk 3 above)
Comment (artlbv) [inline HGCalAnalysis/plugins/HGCalAnalysis.cc]: I've implemented @clelange last suggestion and created another PR to replace this: https://github.com/CMS-HGCAL/reco-ntuples/pull/34
